LinuxAdapter.cs

```diff
@@ -0,0 +1,88 @@
+            SearchRequest searchRequest = new SearchRequest(distinguishedName, filter, SearchScope.Subtree, null);
+            SearchResponse searchResponse = (SearchResponse)connection.SendRequest(searchRequest);
+
+            if (searchResponse.Entries.Count > 0)
+            {
+                var userFound = searchResponse.Entries[0]; //Get the object that was found on ldap
+                string name = userFound.DistinguishedName;
+                var memberof = userFound.Attributes["memberof"]; // You can access ldap Attributes with Attributes property
+
+                foreach (var group in memberof)
+                {
+                    var groupDN = $"{Encoding.UTF8.GetString((byte[])group)}";
+                    var groupCN = groupDN.Split(',')[0].Substring("CN=".Length);
+
+                    if (resolveNestedGroups)
+                    {
+                        GetNestedGroups(connection, context.Principal.Identity, distinguishedName, groupCN);
+                    }
+                    else
+                    {
+                        context.Principal.Identity.AddRole(groupCN);
+                    }
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private static void GetNestedGroups(LdapConnection connection, IIdentity principal, string searchDN, string groupCN)
+        {
+            string filter = $"(&(objectClass=group)(sAMAccountName={groupCN}))"; // This is using ldap search query language, it is looking on the server for someUser
+            SearchRequest searchRequest = new SearchRequest(searchDN, filter, System.DirectoryServices.Protocols.SearchScope.Subtree, null);
+            SearchResponse searchResponse = (SearchResponse)connection.SendRequest(searchRequest);
+
+            if (searchResponse.Entries.Count > 0)
+            {
+                var group = searchResponse.Entries[0]; //Get the object that was found on ldap
+                string name = group.DistinguishedName;
+                principal.AddRole(name);
+
+                var memberof = group.Attributes["memberof"]; // You can access ldap Attributes with Attributes property
+                if (memberof != null)
+                {
+                    foreach (var member in memberof)
+                    {
+                        var groupDN = $"{Encoding.UTF8.GetString((byte[])member)}";
+                        var nestedGroupCN = groupDN.Split(',')[0].Substring("CN=".Length);
+                        GetNestedGroups(connection, principal, searchDN, nestedGroupCN);
+                    }
+                }
+            }
+        }
+
+        private static void AddRole(this IIdentity identity, string role)
+        {
+            var claimsIdentity = identity as ClaimsIdentity;
+            claimsIdentity.AddClaim(new Claim(claimsIdentity.RoleClaimType, role));
```

Comment:
Is this the right way to add the the group information we get from the new LDAP APIs? I tested with `IsInRole` and it works correctly, is that all we care about?

Reply:
ping @blowdart

Reply:
For roles, yup, that's the right way.